ddp-workspace/projects/ddp-dsm-ui/src/app/dashboard-statistics/components/date-range/date-range.component.ts

```diff
@@ -0,0 +1,137 @@
+
+  constructor(private datePipe: DatePipe) {
+  }
+
+  @Input('activeDates') activeDatesOnInit: DateRangeModel;
+
+  @Input('disabled') set disabledState(isDisabled: boolean) {
+    typeof isDisabled === 'boolean' && this.disableOrEnable(isDisabled);
+  }
+
+  @Output() dateChanged = new EventEmitter<DateRangeModel>();
+
+
+  ngOnInit(): void {
+    this.initDateRangeForm();
+    this.listenToValueChangesAndEmit();
+  }
+
+  ngOnDestroy(): void {
+    this.destroyed$.next();
+  }
+
+  public getEntries(object: object): [string, any][] {
+    return Object.entries(object);
+  }
+
+  public get erroredFormControlsEntries(): [string, ValidationErrors][] {
+    return this.getEntries(this.formControls)
+      .reduce(this.filterErroredEntries, []);
+  }
+
+  private initDateRangeForm(): void {
+    if(this.isActiveDateOnInitRangeObject) {
+      this.dateRangeForm.patchValue(this.activeDatesOnInit, {emitEvent: false});
+    }
+  }
+
+  private listenToValueChangesAndEmit(): void {
+    /**
+     * Here is used auditTime operator, due to known bug in Angular material:
+     * Discussion: https://github.com/angular/components/issues/19776
+     */
+    this.dateRangeForm.valueChanges
+      .pipe(auditTime(0), takeUntil(this.destroyed$))
+      .subscribe((dates: DateRangeModel) => this.emitDateChange(dates));
+  }
+
+  private filterErroredEntries(
+    result: [string, ValidationErrors][],
+    [key, value]: [string, AbstractControl]): [string, ValidationErrors][] {
+      value.errors && result.push([key, value.errors]);
+      return result;
+  }
+
+  private emitDateChange(dates: DateRangeModel): void {
+    this.dateRangeForm.valid && this.dateChanged.emit(this.getTransformedDates(dates));
+  }
+
+  private getTransformedDates(dates: DateRangeModel): DateRangeModel {
+    return {
+      startDate: this.datePipe.transform(dates.startDate, Global.DATE_FORMAT),
+      endDate: this.datePipe.transform(dates.endDate, Global.DATE_FORMAT)
+    };
+  }
+
+  private disableOrEnable(disable: boolean): void {
+    disable ? this.disable() : this.enable();
+  }
+
+  private enable(): void {
+    this.startDate.enable({emitEvent: false});
+    this.endDate.enable({emitEvent: false});
+  }
+
+  private disable(): void {
+    this.startDate.disable({emitEvent: false});
+    this.endDate.disable({emitEvent: false});
+  }
+
+  private get generateFormGroup(): FormGroup {
+    return new FormGroup({
+      startDate: new FormControl(null, Validators.required),
+      endDate: new FormControl(null, Validators.required),
+    });
+  }
+
+  private get isActiveDateOnInitRangeObject(): boolean {
+    const requiredProperties: string[] = ['startDate', 'endDate'];
+    /**
+     * Here we are using .some() method on purpose
+     */
+    return this.activeDatesOnInit
+      && this.activeDatesOnInit instanceof Object
+      && requiredProperties.some((prop: string) => this.activeDatesOnInit.hasOwnProperty(prop));
```

Comment:
Are the checks/that `isActiveDateOnInitRangeObject` method needed? 
`DateRangeModel` type should have both fields `startDate` / `endDate`

Reply:
Probably will not, but I still thought, what if this initial data later we want from API, then we might get the wrong value there, or on part missing, as TS won't be able to check at runtime. 
I can remove it, though, as for now, we don't get this info from the backend. Later will think about what approach to take.

P.S. In the case of only one value (either startDate or endDate) is missing, we still display the existing one, that's one I used .some() operator

Reply:
Done

Reply:
If some wrong data is supposed to be from back-end side it is worth use some mapper there
 to align the received data with `DateRangeModel` type.

Reply:
Yes, something like that would be another workaround after integrating API.

Reply:
Yes, something like that would be another workaround after integrating API.